x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/repositories/ClearCcrRestoreSessionAction.java

```diff
@@ -21,34 +22,63 @@
 
 public class ClearCcrRestoreSessionAction extends ActionType<ActionResponse.Empty> {
 
-    public static final ClearCcrRestoreSessionAction INSTANCE = new ClearCcrRestoreSessionAction();
-    public static final String NAME = "internal:admin/ccr/restore/session/clear";
+    public static final ClearCcrRestoreSessionAction INTERNAL_INSTANCE = new ClearCcrRestoreSessionAction();
+    public static final String INTERNAL_NAME = "internal:admin/ccr/restore/session/clear";
+    public static final String NAME = "indices:internal/admin/ccr/restore/session/clear";
+    public static final ClearCcrRestoreSessionAction INSTANCE = new ClearCcrRestoreSessionAction(NAME);
 
     private ClearCcrRestoreSessionAction() {
-        super(NAME, in -> ActionResponse.Empty.INSTANCE);
+        this(INTERNAL_NAME);
     }
 
-    public static class TransportDeleteCcrRestoreSessionAction extends HandledTransportAction<
+    private ClearCcrRestoreSessionAction(String name) {
+        super(name, in -> ActionResponse.Empty.INSTANCE);
+    }
+
+    abstract static class TransportDeleteCcrRestoreSessionAction extends HandledTransportAction<
         ClearCcrRestoreSessionRequest,
         ActionResponse.Empty> {
 
         private final CcrRestoreSourceService ccrRestoreService;
 
-        @Inject
-        public TransportDeleteCcrRestoreSessionAction(
+        private TransportDeleteCcrRestoreSessionAction(
+            String actionName,
             ActionFilters actionFilters,
             TransportService transportService,
             CcrRestoreSourceService ccrRestoreService
         ) {
-            super(NAME, transportService, actionFilters, ClearCcrRestoreSessionRequest::new, ThreadPool.Names.GENERIC);
-            TransportActionProxy.registerProxyAction(transportService, NAME, false, in -> ActionResponse.Empty.INSTANCE);
+            super(actionName, transportService, actionFilters, ClearCcrRestoreSessionRequest::new, ThreadPool.Names.GENERIC);
+            TransportActionProxy.registerProxyAction(transportService, actionName, false, in -> ActionResponse.Empty.INSTANCE);
             this.ccrRestoreService = ccrRestoreService;
         }
 
         @Override
         protected void doExecute(Task task, ClearCcrRestoreSessionRequest request, ActionListener<ActionResponse.Empty> listener) {
+            final ShardId shardId = request.getShardId();
+            // It can be null if the request is sent from an old node with the internal action
+            if (shardId != null) {
```

Comment:
Sanity check: if `shardId` is `null`, for example for `ClearCcrRestoreSessionRequest`, then authorization is guaranteed to fail for the new-style action (`indices:internal/admin/ccr/restore/session/clear`) here:

https://github.com/elastic/elasticsearch/blob/3bd0032bc4b926f2fa21b5e4b1950070819f9935/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java#L149-L151

For the old-style action (`internal:admin/ccr/restore/session/clear`) on the other hand, we won't hit the above IAE because authz won't go through the regular authz flow. 

Is my understanding correct here? If so, I wonder if it makes sense to fail early for new style actions if `shardId` is `null`, e.g., by making `doExecute` or parts of it abstract, and overriding for `InternalTransportAction` and `TransportAction` accordingly. After all the action is guaranteed to fail. 

Apologies if I'm missing something here. I ran out of time getting to the bottom of this today.

Reply:
What you said makes perfect sense! Thanks for the prompt. I was focusing on minimize the changes of different transport actions and forgot about that we could actual leverage the difference. In this case, you are absolutely right that we should fail for `null` in the new `TransportAction` and only have this leniency in the internal one.

This also prompts me to think twice about this [comment](https://github.com/elastic/elasticsearch/pull/95264#discussion_r1174878292). Since we have separate transport actions for different RCS models, we could lift the check up into the transport action so that it does not affect existing model. This is my current preference because conditional check for the `ShardId` here would also make the change not affect existing model. It would be more consistent if both `ShardId` and `fileName` checks either affect the existing model or *not* at all. What do you think?